File: gramatica.cpp
#include <bits/stdc++.h>
#include <readline/history.h>
#include <readline/readline.h>

using namespace std;

void sentence();

int frase_nome(int i);

int frase_nome_p (int i);

int frase_verbal_s (int i);

int frase_verbal_p (int i);

int check_prep (int i);

int check_contracao_s (int j);

int check_contracao_p (int j);

int check_artigo_s (int j);

int check_artigo_p (int j);

int determiner_f(int i);

int determiner_m(int i);

int determiner_p_m(int i);

int determiner_p_f(int i);

int noun_F (int j);

int noun_p_F (int j);

int noun_M (int j);

int noun_p_M(int j);

int prep (int j);

int artigo_f (int j);

int artigo_p_f (int j);

int artigo_m (int j);

int artigo_p_m (int j);

int verb_v (int j);

int verb_ps (int j);

int contracao_m (int j);

int contracao_f (int j);

int contracao_p_m (int j);

int contracao_p_f (int j);

vector<string> x;

set <string> nome_M = {"cacador", "cachorro", "lobo","rio", "mar", "tempo", "sino", "vento", "martelo", "tambor", "tempo","rosto"};
set <string> nome_F = {"cidade","lagrima","mae","menina","noticia", "floresta","porta","vida"};

set <string> nome_p_M = {"cacadores","cachorros","lobos","mares","martelos","tambores","rios","rostos","sinos","ventos"};
set <string> nome_p_F = {"meninas","lagrimas","florestas","maes","vidas","noticias","cidades","portas"};

set <string> verbo_s = {"bate","bateu","corre","correu"};
set <string> verbo_p = {"batem","bateram","batiam","correm","correram","corriam"};

set <string> preposicao = {"com","para"};

set <string> contracao_M = {"no","pelo"};
set <string> contracao_F = {"na","pela"};

set <string> contracao_p_M = {"nos","pelos"};
set <string> contracao_p_F = {"nas","pelas"};

set <string> artigo_F = {"a","A"};
set <string> artigo_M = {"o","O"};

set <string> artigo_p_M = {"os","Os"};
set <string> artigo_p_F = {"as","As"};

void sentence() {
	int i = frase_nome(0);
	if (i<=0) {
		i = frase_nome_p(0);
		if (!(i<=0))
			i = frase_verbal_p(i);
	}
	else if(i<(int)x.size())
		i = frase_verbal_s(i);
	else printf("Erro\n");
	
	if(i!=(int)x.size())
		printf ("Erro\n");
}

int frase_nome(int i) {
	int j = determiner_f(i);
	if (j==-1) {
		j = noun_F(i);
		if (j==-1) {
			j = determiner_m(i);
			if (j==-1){
				j = noun_M(i);
				if (j != -1)
					printf ("sent(noun_prase(noun(%s)),",x[i].c_str());
			}
			else {
				printf ("sent(noun_prase(determiner(%s),",x[i].c_str());
				i=j;
				j = noun_M(j);
				if (j != -1)
					printf ("noun(%s)),",x[i].c_str());
			}
		}
		else
			printf ("sent(noun_prase(noun(%s)),",x[i].c_str());
	}
	else{
		printf ("sent(noun_prase(determiner(%s),",x[i].c_str());
		i=j;
		j = noun_F(j);
		if (j != -1)
			printf ("noun(%s)),",x[i].c_str());
	}
	return j;
}

int frase_nome_p (int i) {
	int j = determiner_p_f(i);
	if (j==-1) {
		j = noun_p_F(i);
		if (j==-1) {
			j = determiner_p_m(i);
			if (j==-1){
				j = noun_p_M(i);
				if (j != -1)
					printf ("sent(noun_prase(noun(%s)),",x[i].c_str());
			}
			else {
				printf ("sent(noun_prase(determiner(%s),",x[i].c_str());
				i=j;
				j = noun_p_M(j);
				if (j != -1)
					printf ("noun(%s)),",x[i].c_str());
			}
		}
		else
			printf ("sent(noun_prase(noun(%s)),",x[i].c_str());
	}
	else {
		printf ("sent(noun_prase(determiner(%s),",x[i].c_str());
		i=j;
		j = noun_p_F(j);
		if (j != -1)
			printf ("noun(%s)),",x[i].c_str());
	}
	return j;
}

int frase_verbal_s (int i) {
	int j = verb_v(i);
	if (j == (int)x.size()) {
		printf ("verbal phrase(verb(%s)))\n",x[i].c_str());
		return j;
	}
	else {
		if (j!=-1) {
			printf ("verbal phrase(verb(%s),", x[i].c_str());
			int i = j;
			j = check_prep (i);
			if (j == -1) {
				j = check_contracao_s(i);
				if (j==-1) {
					check_contracao_p(i);
					if (j == -1) {
						j= check_artigo_s(i);
						if (j==-1)
							j = check_artigo_p (i);
					}
			}
		}
	}
}
	return j;
}

int frase_verbal_p (int i) {	
	int j = verb_ps(i);

	if (j == (int)x.size()){
		printf ("verbal phrase(verb(%s)))\n",x[i].c_str());
		return j;
	}

	else {
		if (j!=-1) {
			printf ("verbal phrase(verb(%s),", x[i].c_str());
			int i = j;
			j = check_prep (i);
			if (j == -1) {
				j = check_contracao_s(i);
				if (j == -1) {
					check_contracao_p(i);
					if (j == -1) {
						j= check_artigo_s(i);
						if (j==-1)
							j = check_artigo_p (i);
					}
			}
		}
	}
}	
	return j;
}

int check_prep (int i) {
	int k = prep (i);
	int j;
	if (k!=-1) {
		printf ("prepostion(%s),",x[i].c_str());
		j = artigo_f (k); 
		if (j==-1) { //se nao for artigo feminino
			j = artigo_m (k);
			if (j == -1) { //se nao for artigo masculino
				j = artigo_p_m (k);
				if (j == -1) { //se nao for artigo plural masculino
					j = artigo_p_f (k);
					if ( j != -1){
						printf ("article(%s),",x[k].c_str());
						i = j;
						j = noun_p_F (j); //verifica se tem nome a seguir
						printf ("noun(%s)))\n",x[i].c_str());
					}
				}
				else{
					printf ("article(%s),",x[k].c_str());
					i = j;
					j = noun_p_M (j); //verifica se tem nome plural masculino a seguir
					printf ("noun(%s)))\n",x[i].c_str());
				}
			}
			else{
				printf ("article(%s),",x[k].c_str());
				i = j;
				j = noun_M (j); //verifica se tem nome masculino
				printf ("noun(%s)))\n",x[i].c_str());
			}
		}
		else{
			printf ("article(%s),",x[k].c_str());
			i = j;
			j = noun_F (j); //verifica se tem nome feminino
			if(j!=-1)
				printf ("noun(%s)))\n",x[i].c_str());
		}
	} 
	else 
		j=k;
	return j;
}

int check_contracao_s (int j) { //contracao no singular
	int i = contracao_m (j);
	if (i == -1) {
		i = contracao_f (j);

		if (i!= -1){
			printf ("contraction(%s),",x[j].c_str());
			j=i;
			i = noun_F (i);
			if(i!=-1)
				printf ("noun(%s)))\n",x[j].c_str());
		}
	}
	else{		
		printf ("contraction(%s),",x[j].c_str());
		j=i;
		i = noun_M (i);
		if(i!=-1)
			printf ("noun(%s)))\n",x[j].c_str());
	}

	return i;
}

int check_contracao_p (int j) { //contracao no plural
	int i = contracao_p_m(j);
	if (i == -1) {
		i = contracao_p_f (j);

		if (i!= -1){
			printf ("contraction(%s),",x[j].c_str());
			j=i;
			i = noun_p_F (i);
			printf ("noun(%s)))\n",x[j].c_str());
		}
	}
	else{
		printf ("contraction(%s),",x[j].c_str());
		j=i;
		i = noun_p_M (i);
		printf ("noun(%s)))\n",x[j].c_str());
	}

	return i;
}

int check_artigo_s (int j) {
	int i = artigo_m (j);
	if (i == -1) {
		i = artigo_f (j);

		if (i!=-1){
			printf ("contraction(%s),",x[j].c_str());
			j=i;
			i = noun_F (i);
			printf ("noun(%s)))\n",x[j].c_str());
		}
	}

	else{
		printf ("article(%s),",x[j].c_str());
		j=i;
		i = noun_M (i);
		printf ("noun(%s)))\n",x[j].c_str());
	}

	return i;
}

int check_artigo_p (int j) {
	int i = artigo_p_m (j);
	if (i == -1) {
		i = artigo_p_f (j);

		if (i!=-1){
			printf ("contraction(%s),",x[j].c_str());
			j=i;
			i = noun_p_F (i);
			printf ("noun(%s)))\n",x[j].c_str());
		}
	}

	else{
		printf ("article(%s),",x[j].c_str());
		j=i;
		i = noun_p_M (i);		
		printf ("noun(%s)))\n",x[j].c_str());
	}

	return i;
}

int determiner_f(int i) {
	if (artigo_F.find(x[i])!=artigo_F.end())
		return i+1;
	else
		return -1;	
}

int determiner_m(int i) {
	if (artigo_M.find(x[i])!=artigo_M.end())
		return i+1;
	else
		return -1;	
}

int determiner_p_m(int i) {
	if (artigo_p_M.find(x[i])!=artigo_p_M.end())
		return i+1;
	else
		return -1;	
}

int determiner_p_f(int i) {
	if (artigo_p_F.find(x[i])!=artigo_p_F.end())
		return i+1;
	else
		return -1;	
}

int noun_F (int j) {
	if (nome_F.find(x[j])!=nome_F.end())
		return j+1;
	else
		return -1;

}

int noun_p_F (int j) {
	if (nome_p_F.find(x[j])!=nome_p_F.end())
		return j+1;
	else
		return -1;

}

int noun_M (int j) {
	if (nome_M.find(x[j])!=nome_M.end())
		return j+1;
	else
		return -1;

}

int noun_p_M(int j) {
	if (nome_p_M.find(x[j])!=nome_p_M.end())
		return j+1;
	else
		return -1;

}

int prep (int j) {
	if (preposicao.find(x[j])!=preposicao.end())
		return j+1;
	else
		return -1;
}

int artigo_f (int j) {
	if (artigo_F.find(x[j])!=artigo_F.end())
		return j+1;
	else
		return -1;
}

int artigo_p_f (int j) {
	if (artigo_p_F.find(x[j])!=artigo_p_F.end())
		return j+1;
	else
		return -1;
}

int artigo_m (int j) {
	if (artigo_M.find(x[j])!=artigo_M.end())
		return j+1;
	else
		return -1;
}

int artigo_p_m (int j) {
	if (artigo_p_M.find(x[j])!=artigo_p_M.end())
		return j+1;
	else
		return -1;
}

int verb_v (int j) {
	if (verbo_s.find(x[j]) != verbo_s.end())
		return j+1;
	else
		return -1;
}

int verb_ps (int j) {
	if (verbo_p.find(x[j]) != verbo_p.end())
		return j+1;
	else
		return -1;
}

int contracao_m (int j) {
	if (contracao_M.find(x[j]) != contracao_M.end())
		return j+1;
	else
		return -1;
}

int contracao_f (int j) {
	if (contracao_F.find(x[j]) != contracao_F.end())
		return j+1;
	else
		return -1;
} 

int contracao_p_m (int j) {
	if (contracao_p_M.find(x[j]) != contracao_p_M.end())
		return j+1;
	else
		return -1;
}

int contracao_p_f (int j) {
	if (contracao_p_F.find(x[j]) != contracao_p_F.end())
		return j+1;
	else
		return -1;
}    


int main () {
	string s;
	for(int i=0;i<28;i++){
		long start_s=clock();
		char *l = readline("String: ");
		string s = l;
		add_history(l);
		istringstream tokenStream(s);
		string token;
			while (getline(tokenStream, token, ' ')) {
				x.push_back(token);
				//printf ("%s\n", token);
			}
		sentence ();
		long stop_s=clock();
    printf("\n\nTotal time thinking: %lf\n",(stop_s-start_s)/double(CLOCKS_PER_SEC)*1000);
    x.clear();
	}
	return 0;
}

File: gramatica.pl
sentence(X) --> frase_nom(X1), frase_verbal(X2), {X=sent(X1,X2)}.
sentence(X) --> frase_nom_p(X1), frase_verbal_p(X2), {X=sent(X1,X2)}.

frase_nom(X) --> determiner_f(X1), noun_f(X2), {X=frase_nom(determiner_f(X1), noun_f(X2))}.
frase_nom(X) --> determiner_m(X1), noun_m(X2), {X=frase_nom(determiner_m(X1), noun_m(X2))}.
frase_nom(X) --> noun_f(X1), {X=frase_nom(noun_f(X1))}.
frase_nom(X) --> noun_m(X1), {X=frase_nom(noun_m(X1))}.

frase_nom_p(X) --> determiner_p_f(X1), noun_p_f(X2), {X=frase_nom_p(determiner_p_f(X1), noun_p_f(X2))}.
frase_nom_p(X) --> determiner_p_m(X1), noun_p_m(X2), {X=frase_nom_p(determiner_p_m(X1), noun_p_m(X2))}.
frase_nom_p(X) --> noun_p_f(X1), {X=frase_nom_p(noun_p_f(X1))}.
frase_nom_p(X) --> noun_p_m(X1), {X=frase_nom_p(noun_p_m(X1))}.

frase_verbal(X) --> verb_n(X1), {X=frase_verbal(verb_n(X1))}.
frase_verbal(X) --> verb_n(X1), detn_m(X2), noun_m(X3), {X=frase_verbal(verb_n(X1), detn_m(X2), noun_m(X3))}.
frase_verbal(X) --> verb_n(X1), detn_f(X2), noun_f(X3), {X=frase_verbal(verb_n(X1), detn_f(X2), noun_f(X3))}.
frase_verbal(X) --> verb_n(X1), detn_p_m(X2), noun_p_m(X3), {X=frase_verbal(verb_n(X1), detn_p_m(X2), noun_p_m(X3))}.
frase_verbal(X) --> verb_n(X1), detn_p_f(X2), noun_p_f(X3), {X=frase_verbal(verb_n(X1), detn_p_f(X2), noun_p_f(X3))}.
frase_verbal(X) --> verb_v(X1), {X=frase_verbal(verb_v(X1))}.
frase_verbal(X) --> verb_v(X1), detv_m(X2), noun_m(X3), {X=frase_verbal(verb_v(X1), detv_m(X2), noun_m(X3))}.
frase_verbal(X) --> verb_v(X1), detv_f(X2), noun_f(X3), {X=frase_verbal(verb_v(X1), detv_f(X2), noun_f(X3))}.
frase_verbal(X) --> verb_v(X1), detv_p_m(X2), noun_p_m(X3), {X=frase_verbal(verb_v(X1), detv_p_m(X2), noun_p_m(X3))}.
frase_verbal(X) --> verb_v(X1), detv_p_f(X2), noun_p_f(X3), {X=frase_verbal(verb_v(X1), detv_p_f(X2), noun_p_f(X3))}.

frase_verbal_p(X) --> verb_p_n(X1), {X=frase_verbal_p(verb_p_n(X1))}.
frase_verbal_p(X) --> verb_p_n(X1), detn_p_m(X2), noun_p_m(X3), {X=frase_verbal_p(verb_p_n(X1), detn_p_m(X2), noun_p_m(X3))}.
frase_verbal_p(X) --> verb_p_n(X1), detn_p_f(X2), noun_p_f(X3), {X=frase_verbal_p(verb_p_n(X1), detn_p_f(X2), noun_p_f(X3))}.
frase_verbal_p(X) --> verb_p_n(X1), detn_m(X2), noun_m(X3), {X=frase_verbal_p(verb_p_n(X1), detn_m(X2), noun_m(X3))}.
frase_verbal_p(X) --> verb_p_n(X1), detn_f(X2), noun_f(X3), {X=frase_verbal_p(verb_p_n(X1), detn_f(X2), noun_f(X3))}.
frase_verbal_p(X) --> verb_p_v(X1), {X=frase_verbal_p(verb_p_v(X1))}.
frase_verbal_p(X) --> verb_p_v(X1), detv_p_m(X2), noun_p_m(X3), {X=frase_verbal_p(verb_p_v(X1), detv_p_m(X2), noun_p_m(X3))}.
frase_verbal_p(X) --> verb_p_v(X1), detv_p_f(X2), noun_p_f(X3), {X=frase_verbal_p(verb_p_v(X1), detv_p_f(X2), noun_p_f(X3))}.
frase_verbal_p(X) --> verb_p_v(X1), detv_m(X2), noun_m(X3), {X=frase_verbal_p(verb_p_v(X1), detv_m(X2), noun_m(X3))}.
frase_verbal_p(X) --> verb_p_v(X1), detv_f(X2), noun_f(X3), {X=frase_verbal_p(verb_p_v(X1), detv_f(X2), noun_f(X3))}.


determiner_m(X)   --> [o], {X=o}; ['O'], {X='O'}.
determiner_f(X)   --> [a], {X=a}; ['A'], {X='A'}.

determiner_p_m(X) --> [os], {X=os}; ['Os'], {X='Os'}.
determiner_p_f(X) --> [as], {X=as}; ['As'], {X='As'}.

noun_m(X)    --> [tempo], {X=tempo}; [cacador], {X=cacador}; [rosto], {X=rosto}; [rio], {X=rio}; [mar], {X=mar}; [martelo], {X=martelo}; [cachorro], {X=cachorro}; [tambor], {X=tambor}; [sino], {X=sino}; [lobo], {X=lobo}; [vento], {X=vento}.
noun_p_m(X)  --> [cacadores], {X=cacadores}; [rostos], {X=rostos}; [rios], {X=rios}; [mares], {X=mares}; [martelos], {X=martelos}; [cachorros], {X=cachorros}; [tambores], {X=tambores}; [sinos], {X=sinos}; [lobos], {X=lobos}; [ventos], {X=ventos}.

noun_f(X)    --> [menina], {X=menina}; [floresta], {X=floresta}; [mae], {X=mae}; [vida], {X=vida}; [noticia], {X=noticia}; [cidade], {X=cidade}; [lagrima], {X=lagrima}; [porta], {X=porta}.
noun_p_f(X)  --> [meninas], {X=meninas}; [florestas], {X=florestas}; [maes], {X=maes}; [vidas], {X=vidas}; [noticias], {X=noticias}; [cidades], {X=cidades}; [lagrimas], {X=lagrimas}; [portas], {X=portas}.

verb_n(X)    --> [bate], {X=bate}; [bateu], {X=bateu}.
verb_p_n(X)  --> [bateram], {X=bateram}; [batiam], {X=batiam}; [batem], {X=batem}.

verb_v(X)    --> [corre], {X=corre}; [correu], {X=correu}.
verb_p_v(X)  --> [correram], {X=correram}; [corriam], {X=corriam}; [correm], {X=correm}.

detn_m(X)    --> [no], {X=no}; [o], {X=o}.
detn_p_m(X)  --> [nos], {X=nos}; [os], {X=os}.

detn_f(X)    --> [na], {X=na}; [a], {X=a}.
detn_p_f(X)  --> [nas], {X=nas}; [as], {X=a}.

detv_m(X)    --> [com], [o], {X=[com,o]}; [pelo], {X=pelo}; [para], [o], {X=[para,o]}.
detv_p_m(X)  --> [com], [os], {X=[com,os]}; [pelos], {X=pelos}; [para], [os], {X=[para,os]}.

detv_f(X)    --> [com], [a], {X=[com,as]}; [pela], {X=pela}; [para], [a], {X=[para,a]}.
detv_p_f(X)  --> [com], [as], {X=[com,as]}; [pelas], {X=pelas}; [para], [as], {X=[para,as]}.

